File: 1100-connecting-cities-with-minimum-cost/connecting-cities-with-minimum-cost.py
class Solution:
    def minimumCost(self, n: int, connections: List[List[int]]) -> int:
        def find(x):
            if p[x] != x:
                p[x] = find(p[x])
            return p[x]

        def union(a, b):
            ap = find(a)
            bp = find(b)
            p[ap] = bp

        p = [i for i in range(n + 1)]
        connections.sort(key=lambda x:x[-1])
        total = 0

        for u, v, c in connections:
            if find(u) != find(v):
                union(u, v)
                total += c
        
        roots = set()
        for i in range(1, n + 1):
            roots.add(find(i))

        if len(roots) == 1:
            return total
        return -1

File: 1100-connecting-cities-with-minimum-cost/connecting-cities-with-minimum-cost.js
/**
 * @param {number} n
 * @param {number[][]} connections
 * @return {number}
 */
var minimumCost = function(n, connections) {
    const find = (x) => {
        if (p[x] !== x) {
            p[x] = find(p[x]);
        }
        return p[x]
    }

    const union = (a, b) => {
        const ap = p[a];
        const bp = p[b];

        p[ap] = bp;
    }

    connections.sort((a, b) => a[2] - b[2]);

    const p = Array.from({length : n + 1}, (_, i) => i);
    let cost = 0;
    for (let i = 0; i < connections.length; i++) {
        const [u, v, c] = connections[i];

        if (find(u) !== find(v)) {
            cost += c;
            union(u, v);
        }
    }

    const area = new Set();

    for (let i = 1; i <= n; i++) {
        area.add(find(i))
    }
    return area.size === 1 ? cost : -1;
};